casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -30,7 +31,6 @@ import coop.rchain.shared._
   @param equivocationsTracker: Used to keep track of when other validators detect the equivocation consisting of the base block at the sequence number identified by the (validator, base equivocation sequence number) pair of each EquivocationRecord.
   */
 final case class CasperState(
-    seenBlockHashes: Set[BlockHash] = Set.empty[BlockHash],
```

Comment:
AFAIU from our discussion with @KentShikama, this field is not exactly equivalent to `dag.contains` because `seenBlockHashes` also includes blocks which were seen but were not added just yet (i.e. the blocks from `blockBuffer`). I believe that is why it was added in the first place. @lukasz-golebiewski @KentShikama please correct me if I am wrong. The tests didn't catch that the current implementation is buggy which makes me worry.

So the correct fix here would be to replace `seenBlockHashes` with a set of block hashes from `blockBuffer` which is king of what you are already doing in the second checkpoint.